extension/amqp/listener.go

```diff
@@ -0,0 +1,132 @@
+package amqp
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/hellofresh/goengine"
+	"github.com/hellofresh/goengine/driver/sql"
+	"github.com/mailru/easyjson"
+	"github.com/streadway/amqp"
+)
+
+// Ensure Listener implements sql.Listener
+var _ sql.Listener = &Listener{}
+
+type (
+	// Consume returns a channel of amqp.Delivery's and a related closer or an error
+	Consume func() (io.Closer, <-chan amqp.Delivery, error)
+
+	// Listener consumes messages from an queue
+	Listener struct {
+		consume              Consume
+		minReconnectInterval time.Duration
+		maxReconnectInterval time.Duration
+		logger               goengine.Logger
+	}
+)
+
+// NewListener returns a new Listener
+func NewListener(
+	consume Consume,
+	minReconnectInterval time.Duration,
+	maxReconnectInterval time.Duration,
+	logger goengine.Logger,
+) (*Listener, error) {
+	switch {
+	case consume == nil:
+		return nil, goengine.InvalidArgumentError("consume")
+	}
+
+	if logger == nil {
+		logger = goengine.NopLogger
+	}
+
+	return &Listener{
+		consume:              consume,
+		minReconnectInterval: minReconnectInterval,
+		maxReconnectInterval: maxReconnectInterval,
+		logger:               logger,
+	}, nil
+}
+
+// Listen receives messages from a queue, transforms them into a sql.ProjectionNotification and calls the trigger
+func (l *Listener) Listen(ctx context.Context, trigger sql.ProjectionTrigger) error {
+	var nextReconnect time.Time
+	reconnectInterval := l.minReconnectInterval
+	for {
+		select {
+		case <-ctx.Done():
+			return context.Canceled
+		default:
+		}
+
+		conn, deliveries, err := l.consume()
+		if err != nil {
+			l.logger.Error("failed to start consuming amqp messages", func(entry goengine.LoggerEntry) {
+				entry.Error(err)
+				entry.String("reconnect_in", reconnectInterval.String())
+			})
+
+			time.Sleep(reconnectInterval)
+			reconnectInterval *= 2
+			if reconnectInterval > l.maxReconnectInterval {
+				reconnectInterval = l.maxReconnectInterval
+			}
+			continue
+		}
+		reconnectInterval = l.minReconnectInterval
+		nextReconnect = time.Now().Add(reconnectInterval)
+
+		l.consumeMessages(ctx, conn, deliveries, trigger)
+
+		select {
+		case <-ctx.Done():
+			return context.Canceled
+		default:
+			time.Sleep(time.Until(nextReconnect))
+		}
+	}
+}
+
+func (l *Listener) consumeMessages(ctx context.Context, conn io.Closer, deliveries <-chan amqp.Delivery, trigger sql.ProjectionTrigger) {
```

Comment:
The `Listener` as designed cannot safely shutdown. Yes, if you close the context, the Listener will eventually shutdown (at some point it will randomly choose the `<-ctx.Done` over `<-deliveries`), but it will have potentially left `amqp.Delivery`s waiting [inside of the the `streadway/amqp` package](https://github.com/streadway/amqp/blob/e5adc2ada8b8efff032bf61173a233d143e9318e/consumers.go#L59) to be sent over the `chan amqp.Delivery` and thus unhandled.

Combined with `autoAck` these messages would also be completely lost during shutdown, and never redelivered to a new consumer, because they were acked at delivery from the server. And even more worse, since the `Qos` value will be ignored, this could drop a large amount of `amqp.Delivery`s.